### ISSUE TITLE
FK_DB055AF32B4057C1: Integrity constraint violation: 1451 Cannot delete or update a parent row

### ISSUE THREAD

**Issue by Digi92** [mosparo/mosparo#97]
Hello,
i have mosparo version 0.3.20 on a small site in testing.
Now it has come to a database error, whereby the captach has displayed an error message.
I could also trigger the error with the command "mosparo:cleanup-database".

I fixed the error message by changing the "ON DELETE" of the foreign key "FK_DB055AF32B4057C1" of the "submission" table from "RESTRICT" to "CASCADE".
So the DELETE statement could be executed but i am not sure if this was the correct solution?

Error message from log:
```
[2023-06-21T13:21:32.072463+02:00] request.CRITICAL: Uncaught PHP Exception Doctrine\DBAL\Exception\ForeignKeyConstraintViolationException: "An exception occurred while executing 'DELETE FROM submit_token WHERE (created_at < ? AND submission_id IS NULL) AND (submit_token.project_id = 2)' with params ["2023-06-20 13:21:32"]:  SQLSTATE[23000]: Integrity constraint violation: 1451 Cannot delete or update a parent row: a foreign key constraint fails (`captcha_api`.`submission`, CONSTRAINT `FK_DB055AF32B4057C1` FOREIGN KEY (`submit_token_id`) REFERENCES `submit_token` (`id`))" at /vendor/doctrine/dbal/lib/Doctrine/DBAL/Driver/AbstractMySQLDriver.php line 68 {"exception":"[object] (Doctrine\\DBAL\\Exception\\ForeignKeyConstraintViolationException(code: 0): An exception occurred while executing 'DELETE FROM submit_token WHERE (created_at < ? AND submission_id IS NULL) AND (submit_token.project_id = 2)' with params [\"2023-06-20 13:21:32\"]:\n\nSQLSTATE[23000]: Integrity constraint violation: 1451 Cannot delete or update a parent row: a foreign key constraint fails (`captcha_api`.`submission`, CONSTRAINT `FK_DB055AF32B4057C1` FOREIGN KEY (`submit_token_id`) REFERENCES `submit_token` (`id`)) at /vendor/doctrine/dbal/lib/Doctrine/DBAL/Driver/AbstractMySQLDriver.php:68)\n[previous exception] [object] (Doctrine\\DBAL\\Driver\\PDO\\Exception(code: 23000): SQLSTATE[23000]: Integrity constraint violation: 1451 Cannot delete or update a parent row: a foreign key constraint fails (`captcha_api`.`submission`, CONSTRAINT `FK_DB055AF32B4057C1` FOREIGN KEY (`submit_token_id`) REFERENCES `submit_token` (`id`)) at /vendor/doctrine/dbal/lib/Doctrine/DBAL/Driver/PDO/Exception.php:18)\n[previous exception] [object] (PDOException(code: 23000): SQLSTATE[23000]: Integrity constraint violation: 1451 Cannot delete or update a parent row: a foreign key constraint fails (`captcha_api`.`submission`, CONSTRAINT `FK_DB055AF32B4057C1` FOREIGN KEY (`submit_token_id`) REFERENCES `submit_token` (`id`)) at /vendor/doctrine/dbal/lib/Doctrine/DBAL/Driver/PDOStatement.php:117)"} []

```

**Comment by zepich** [comment]
Hi @Digi92 

Thank you very much for your report!

As far as I can tell, it is generally a logic bug between the two entities. `Submission` has a `ManyToOne` relation with `SubmitToken`, while the `SubmitToken` has an OneToOne relation with `Submission`. Normally, both entities should have the same relation (`OneToOne` + `OneToOne` or `ManyToOne + OneToMany`).

I guess the message you see is because you had multiple submissions with the same submit token. A submit token can be used for numerous submissions until the form is validated. The cleanup command (`CleanupHelper`) prepared the entities correctly but removed only the latest submission. The firstly created submissions with the same submit token are still there, generating this issue. 

The correct solution is to solve the issue with the connection. My initial idea was probably to assign only the latest submission with the submit token. But since the submit token contains the IP address (and other information), we probably switched that to a `ManyToOne` on the submission side without updating the cleanup process (to show the IP address and other information with every submission).

I think your fix is not the best solution. It will help clean up your table, of course. But I guess you will have a problem sooner or later because the submit token for a submission is missing - especially because of a bug that we fixed in v0.4.0 with the cleanup process deleting spam and invalid submissions too early (#84, https://github.com/mosparo/mosparo/commit/827217f8a6a10ab5d99f503d1bbe650a4736c13a). The best solution is to adjust the relation between the two entities and to adjust the cleanup helper to work correctly.

Is there a specific reason you haven't upgraded to v0.3.20 yet? Or hadn't time yet or want to wait for the first minor update? We didn't plan to release a new version for 0.3, but if there is a reason for you not to upgrade to 0.4, we have to think about that again.

I will investigate the problem and prepare the correct fix for it.

**Comment by zepich** [comment]
I took some time and investigated the issue further.

**Relationship:**
The "incorrect" relationship with OneToOne and ManyToOne is correct, but the property name is not good. The submit token needs to know the last submission because only the last submission will be verified if the verification process happens. The last submission must be stored on the submit token in mosparo, not in the frontend. Otherwise, a bad person could verify an older submission.

But the property name on the submit token should not be `submission` but `lastSubmission` (or something like that).

**The cleanup functionality:**
At this point, I'm not able to reproduce your issue. In general, submissions get deleted before the submit token can be deleted. Valid (and in 0.4, also invalid) submissions get deleted after 14 days, and invalid submissions after 24 hours (in 0.3). The submit tokens get deleted if no submission is set and after 24 hours.

So even if you use a submit token only for invalid submissions, the invalid submissions should be deleted first.

I will invest more time in understanding this issue. 

Have you tested anything unusual with the frontend API which could lead to this issue?

**Comment by Digi92** [comment]
Thanks for the quick answer.

I had not yet updated it to 0.4 because I had not found the time.
However, since the bug from#84 seemed important and the fix was a one-liner, I had already installed it by hand last week. So the bug should have nothing to do with the problem.

There was no unusual testing against the frontend API but since, as mentioned, it is on a small site in real world use, site visitors could have done something unexpected here.
However, there is at least no indication of this in the log.
The only thing I notice is that before the error occurred, there was a submission which is not noticeable in the spam check but was invalid in the validation. The content of the form was not suspicious.
But after that, there were also other valid form submissions before the error occurred.

**Comment by zepich** [comment]
Hi @Digi92 

Okay, thank you very much for your information.

I will investigate the issue further and add a fix with v0.4.1.

**Comment by zepich** [comment]
Okay, so after another round of testing, I can reproduce the issue. The "funny" story: the fix from v0.4.0 created the issue (well, at least the fix from 0.4.0 made the problem visible).

1. When You open the form, a submit token is requested from mosparo
2. You fill out the form and enter some bad content
3. You click the mosparo box, your form data will be checked, and the submission is stored in the database with the ID of the submit token.
4. You adjust your content and click the mosparo box again. mosparo again checks your submission, and everything is good. The new submission is also stored with the ID of the submit token.
5. Now, **do not** submit the form. 

After 24 hours, the last valid submission will be deleted because it was not verified (not submitted). But the first submission, detected as spam, will not be deleted because of the bug fix. It has to stay in the database for another 13 days.

Before v0.4.0, the spam submission got deleted after 24 hours, so the problem was not visible.

I first tested it with a v0.3.20 since you initially wrote that you're using v0.3.20, where I couldn't reproduce the issue.

I will work on a fix and release it as v0.4.1. 

Thank you very much for your reporting and testing!

**Comment by zepich** [comment]
v0.4.1 has been released, which contains the promised fix for this issue (and I renamed the misleading property name).
